torch/testing/_asserts.py

```diff
@@ -649,7 +666,7 @@ def assert_close(
         To assert that the values in tensors are bitwise equal, use :func:`assert_equal` instead.
     """
     check_tensors = functools.partial(
-        _check_tensors_close,
+        _cast_inputs(_check_tensors_close),
```

Comment:
I'm not in love with the idea that scalars, lists, arrays and tensors can be "close." 

Arrays and tensors make sense to me (although we should be sure that if a NumPy array is on the CPU and a tensor is on a CUDA device then check_device must be False), but if we want to cast other types to tensors maybe we need a different function or flag.

Note that in general this approach won't work for supporting scalar <-> tensor or list <-> tensor comparisons because it will also require check_device=False to work with CUDA tensors (as mentioned above). So a more permissive function that also doesn't check device would seem appropriate for this case.

Reply:
(Not that I think this case is necessarily worth supporting. I generally wouldn't want engineers to use a more permissive cross-type version because it'd be harder to verify that the operation was returning the correct type of object.)

Reply:
My idea behind all this was to support three basic types:

1. `torch.Tensor`,
2. `np.ndarray`, and
3. python scalars (`int`, `float`, and later `complex`)

They are simply cast into a tensor to use the same machinery for everything. Furthermore, I want to support sequences and mappings of any of the valid types, which will be checked elementwise.

> I'm not in love with the idea that scalars, lists, arrays and tensors can be "close."
>
> Arrays and tensors make sense to me [...], but if we want to cast other types to tensors maybe we need a different function or flag.

- **scalars**: Not sure why a scalar shouldn't be able to be close to a tensor. What is wrong with checking `torch.testing.assert_close(torch.tensor(1.0), 1.0)`?
- **lists**: I assume you mean lists of scalars, correct? This is basically a byproduct, because I didn't enforce only the three types above. So right now you could do `torch.testing.assert_close([1.0, 2.0], torch.tensor([1.0, 2.0]))`. If we don't want that, bailing out in such a scenario can be added.

> although we should be sure that if a NumPy array is on the CPU and a tensor is on a CUDA device then check_device must be False

I haven't actually thought about that. The same would be needed for `check_dtype` if we allow scalars. We could have them as `Optional[bool] = None` and set this depending on the situation.

Thoughts?